src/main/java/org/openrewrite/kotlin/internal/KotlinParserVisitor.java

```diff
@@ -86,22 +87,47 @@ public class KotlinParserVisitor extends FirDefaultVisitor<J, ExecutionContext>
     private final FirSession firSession;
     private int cursor;
 
+    private final Map<Integer, ASTNode> nodes = new HashMap<>();
+
     // Associate top-level function and property declarations to the file.
     @Nullable
     private FirFile currentFile;
 
-    private static final Pattern whitespaceSuffixPattern = Pattern.compile("\\s*[^\\s]+(\\s*)");
-
-    public KotlinParserVisitor(Path sourcePath, @Nullable FileAttributes fileAttributes, EncodingDetectingInputStream source, List<NamedStyles> styles, JavaTypeCache typeCache, FirSession firSession, ExecutionContext ctx) {
-        this.sourcePath = sourcePath;
-        this.fileAttributes = fileAttributes;
-        this.source = source.readFully();
-        this.charset = source.getCharset();
-        this.charsetBomMarked = source.isCharsetBomMarked();
+    public KotlinParserVisitor(KotlinSource kotlinSource, @Nullable Path relativeTo, List<NamedStyles> styles, JavaTypeCache typeCache, FirSession firSession, ExecutionContext ctx) {
+        this.sourcePath = kotlinSource.getInput().getRelativePath(relativeTo);
+        this.fileAttributes = kotlinSource.getInput().getFileAttributes();
+        EncodingDetectingInputStream is = kotlinSource.getInput().getSource(ctx);
+        this.source = is.readFully();
+        this.charset = is.getCharset();
+        this.charsetBomMarked = is.isCharsetBomMarked();
         this.styles = styles;
         this.typeMapping = new KotlinTypeMapping(typeCache, firSession);
         this.ctx = ctx;
         this.firSession = firSession;
+        init(kotlinSource.getKtFile());
+    }
+
+    private void init(KtFile ktFile) {
+        PsiElementVisitor v = new PsiElementVisitor() {
+            @Override
+            public void visitElement(@NotNull PsiElement element) {
+                if (!(element instanceof KtFile) && element.getTextLength() != 0) {
+                    // Set the node at the cursor to the last visited element.
+                    nodes.put(element.getTextRange().getStartOffset(), element.getNode());
+                    assert source.substring(element.getTextRange().getStartOffset(), element.getTextRange().getEndOffset()).equals(element.getText());
+                }
+
+                for (PsiElement child : element.getChildren()) {
+                    if (child != null) {
+                        visitElement(child);
+                    }
+                }
+                if (element.getNextSibling() != null) {
```

Comment:
I think line 125-127 is not necessary and redundant for the current solution, since `NextSibling` is a child of the parent element visited above. Removing these can improve the performance of visiting, otherwise the visiting for a large file will be heavy

Reply:
Nice catch -- updated to prevent duplicate visits.